### ISSUE TITLE
Handle new `caller_location` intrinsic the right way

### ISSUE THREAD

**Issue by adpaco-aws** [model-checking/rmc#374]
Rebase from #346 found a new `caller_location` intrinsic in the process. In order to unblock, we codegen this intrinsic as an `SKIP` statement:
```rust
"caller_location" => Stmt::skip(loc),
```
But we can probably codegen it the same way Rust does.

**Comment by celinval** [comment]
This is a nice explanation on what this does: https://hegdenu.net/posts/track-caller/

**Comment by adpaco-aws** [comment]
Thanks for the update, @celinval !

If adding support for `caller_location` would allow us to avoid panic messages such as ``thread 'main' panicked at 'called `Option::unwrap()` on a `None` value'``, that'd be a major improvement in how we report such failures.

One thing that I don't see mentioned there is the relation to panic strategies. This isn't mentioned in the [RFC](https://github.com/rust-lang/rfcs/pull/2091) neither, so I guess it's independent.

**Comment by celinval** [comment]
I was wondering if it would help with: https://github.com/model-checking/kani/issues/1740

**Comment by celinval** [comment]
For the current Rust implementation, the compiler will add an extra argument to functions that contain `#[track_caller]`. Unfortunately, this operation is done in the codegen stage, instead of a MIR pass. :(

My guess is that MIR passes can't really change the function signature, so they push it down to the codegen stage. I'm tempted to just encode this as a global variable for now.

@adpaco-aws I don't think CBMC really supports formatting error messages, so I'm not sure implementing this intrinsic will make much of a difference on how we report failures.

In order to fix that, we would have to either inline these functions or generate a different copy per location they are invoked from.

**Comment by celinval** [comment]
Implementation notes for myself:

1. To build a constant that represents a location, use query `const_caller_location` [[1]](https://github.com/rust-lang/rust/blob/0ab38e95bb1cbf0bd038d359bdecbfa501f003a7/compiler/rustc_middle/src/query/mod.rs#L1097).
2. To get the type of the caller location, use function: `pub fn caller_location_ty(self) -> Ty<'tcx>` [[2]](https://github.com/rust-lang/rust/blob/0ab38e95bb1cbf0bd038d359bdecbfa501f003a7/compiler/rustc_middle/src/ty/context.rs#L1156)
3. We can check if a function requires caller location using `requires_caller_location` [[3]](https://github.com/rust-lang/rust/blob/0ab38e95bb1cbf0bd038d359bdecbfa501f003a7/compiler/rustc_middle/src/ty/instance.rs#L257)
4. If we are passing the caller location as an argument, we will need to create a ReifyShim for cases that require dynamic dispatch.

**Comment by joshlf** [comment]
Ran into this when implementing zstd decompression (RFC 8878). Should I consider the `proof` function to be an incomplete proof as a result?

Source:

```rust
// https://datatracker.ietf.org/doc/html/rfc8878#name-window-descriptor
#[repr(transparent)]
struct WindowDescriptor(u8);

impl WindowDescriptor {
    fn window_size(self) -> u64 {
        #[cfg(kani)]
        #[kani::proof]
        fn proof() {
            let desc = WindowDescriptor(kani::any());
            let size = desc.window_size();
            assert!(size <= (1u64 << 41) + 7*(1u64 << 38));
        }

        let exponent = self.0 >> 3;
        let mantissa = self.0 & 0b111;

        // SAFETY: Proven sound by Kani proof.
        let window_log = unsafe { 10u32.unchecked_add(u32::from(exponent)) };
        let window_base = 1u64 << window_log;
        // SAFETY: Proven sound by Kani proof.
        let window_add = unsafe { (window_base / 8).unchecked_mul(u64::from(mantissa)) };
        // SAFETY: Proven sound by Kani proof.
        unsafe { window_base.unchecked_add(window_add) }
    }
}
```

**Comment by zhassan-aws** [comment]
> Should I consider the proof function to be an incomplete proof as a result?

No, it shouldn't. Verification would fail if the intrinsic is reachable from the proof harness.

**Comment by celinval** [comment]
This CR might be of interest here. It adds support to compute caller location using  rustc_public APIs:  https://github.com/rust-lang/rust/pull/159204
